File: web/app/api/mcp/tools/search-community-fixes.ts
import { db, errorPatterns, communityFixes } from "@/lib/db";
import { sql, desc } from "drizzle-orm";
import type { McpUser } from "../auth";

export async function execute(
  args: Record<string, unknown>,
  _user: McpUser
): Promise<string> {
  const query = args.query as string;
  if (!query) return "Error: query is required.";

  const limit = Math.min(Number(args.limit) || 5, 20);

  // Trigram similarity search on pattern_text
  const results = await db.execute(sql`
    SELECT
      p.id AS pattern_id,
      p.pattern_text,
      p.category,
      p.framework,
      p.language,
      p.occurrence_count,
      cf.id AS fix_id,
      cf.fix_approach,
      cf.fix_description,
      cf.avg_confidence,
      cf.success_count,
      cf.total_applications
    FROM error_patterns p
    LEFT JOIN community_fixes cf ON cf.pattern_id = p.id
    WHERE similarity(p.pattern_text, ${query}) > 0.1
    ORDER BY similarity(p.pattern_text, ${query}) DESC, cf.success_count DESC
    LIMIT ${limit * 3}
  `);

  const rows = results.rows as Array<Record<string, unknown>>;
  if (rows.length === 0) return `No community fixes found for: ${query}`;

  // Group by pattern
  const patterns = new Map<string, { pattern: Record<string, unknown>; fixes: Array<Record<string, unknown>> }>();

  for (const r of rows) {
    const pid = r.pattern_id as string;
    if (!patterns.has(pid)) {
      patterns.set(pid, {
        pattern: {
          patternText: r.pattern_text,
          category: r.category,
          framework: r.framework,
          language: r.language,
          occurrences: r.occurrence_count,
        },
        fixes: [],
      });
    }
    if (r.fix_id) {
      patterns.get(pid)!.fixes.push({
        approach: r.fix_approach,
        description: r.fix_description,
        confidence: r.avg_confidence,
        successCount: r.success_count,
        total: r.total_applications,
      });
    }
  }

  let out = `Community fixes for "${query}"\n${"=".repeat(50)}\n\n`;
  let i = 0;
  for (const [, { pattern, fixes }] of patterns) {
    if (i >= limit) break;
    i++;

    out += `${i}. Pattern: ${pattern.patternText}\n`;
    out += `   Category: ${pattern.category}`;
    if (pattern.framework) out += ` | Framework: ${pattern.framework}`;
    if (pattern.language) out += ` | Language: ${pattern.language}`;
    out += ` | Seen ${pattern.occurrences}x\n`;

    for (const fix of fixes.slice(0, 3)) {
      const total = Number(fix.total) || 0;
      const success = Number(fix.successCount) || 0;
      const pct = total > 0 ? `${Math.round((success / total) * 100)}% success (${success}/${total})` : "new fix";
      out += `   Fix: ${fix.approach} — ${fix.description}\n`;
      out += `        Confidence: ${fix.confidence}% | ${pct}\n`;
    }
    out += "\n";
  }

  return out;
}

File: cli/src/mcp/tools/search_community_fixes.rs
use serde_json::Value;

use crate::config;

/// Search the InariWatch community fix network for known fixes matching a query.
/// Calls GET {fix_replay_url}/api/patterns/search?q=<query>
pub async fn execute(args: &Value) -> anyhow::Result<String> {
    let query = args["query"]
        .as_str()
        .ok_or_else(|| anyhow::anyhow!("query is required"))?;
    let limit = args["limit"].as_u64().unwrap_or(5).min(20) as usize;

    let cfg = config::load()?;
    let base_url = cfg
        .global
        .fix_replay_url
        .as_deref()
        .unwrap_or("https://app.inariwatch.com");

    let encoded_q = urlencoding::encode(query);
    let url = format!(
        "{}/api/patterns/search?q={}",
        base_url.trim_end_matches('/'),
        encoded_q
    );

    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()?;

    let resp = client.get(&url).send().await?;
    if !resp.status().is_success() {
        return Ok(format!(
            "Community fix network unavailable (HTTP {}). Is fix_replay_url configured correctly?",
            resp.status()
        ));
    }

    let body: Value = resp.json().await?;

    let matches = match body["matches"].as_array() {
        Some(m) if !m.is_empty() => m,
        _ => return Ok(format!("No community fixes found for: {}", query)),
    };

    let mut out = format!(
        "Community fixes for \"{}\"\n{}\n\n",
        query,
        "=".repeat(50)
    );

    for (i, m) in matches.iter().take(limit).enumerate() {
        let pattern_text = m["pattern"]["patternText"].as_str().unwrap_or("Unknown pattern");
        let category = m["pattern"]["category"].as_str().unwrap_or("unknown");
        let framework = m["pattern"]["framework"].as_str().unwrap_or("");
        let language = m["pattern"]["language"].as_str().unwrap_or("");

        out.push_str(&format!(
            "{}. Pattern: {}\n   Category: {}",
            i + 1,
            pattern_text,
            category
        ));
        if !framework.is_empty() {
            out.push_str(&format!(" | Framework: {}", framework));
        }
        if !language.is_empty() {
            out.push_str(&format!(" | Language: {}", language));
        }
        out.push('\n');

        if let Some(fixes) = m["fixes"].as_array() {
            for fix in fixes.iter().take(3) {
                let approach = fix["fixApproach"].as_str().unwrap_or("");
                let description = fix["fixDescription"].as_str().unwrap_or("");
                let success = fix["successCount"].as_u64().unwrap_or(0);
                let total = fix["totalApplications"].as_u64().unwrap_or(0);
                let confidence = fix["avgConfidence"].as_u64().unwrap_or(0);
                let success_pct = if total > 0 {
                    format!("{}% success ({}/{})", success * 100 / total, success, total)
                } else {
                    "new fix".to_string()
                };

                out.push_str(&format!(
                    "   Fix: [{}] {} — {}\n        Confidence: {}% | {}\n",
                    approach, description, "", confidence, success_pct
                ));
            }
        }
        out.push('\n');
    }

    Ok(out)
}
